### PR TITLE
enh: add dataset exampels

### PR DESCRIPTION
**Pull request by blasscoc**
Adds an annoted example for the dataset class.

### CODE REVIEW

File: examples/dataset_example/src/dataset_example.cc [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+  /// handled by returning a "Result" object; the result can be tested for
+  /// "ok()", there is also a convenient Tensorstore macro for doing assigment
+  /// while handling errors.
+  MDIO_ASSIGN_OR_RETURN(auto dataset, dataset_result)
+
+  /// The dataset represent data on disk, the object holds only minimal state.
+  /// Its memory footprint is small.
+  std::cout << dataset << "\n" << std::endl;
+
+  /// Variable type is discoverable at runtime, initially the variable will have
```
Comment (BrianMichell) [inline examples/dataset_example/src/dataset_example.cc:75]: I might consider using the `at(std::string name)` method over `get<void>(std::string name);` here. It provides a lighter weight method to query the Variable properties' as demonstrated.

File: examples/dataset_example/src/dataset_example.cc [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+      variableObject.get_store().domain()[0].interval().exclusive_max();
+
+  /// Reads in tensorstore are asynchronous, we use the "result()" method to
+  /// await the read.
+  // MDIO_ASSIGN_OR_RETURN(
+  //   auto data, tensorstore::Read(variableObject.get_store()).result()
+  // )
+  MDIO_ASSIGN_OR_RETURN(auto data, variableObject.Read().result())
+  auto d1 = data.get_data_accessor();
+  /// The read returns a SharedArray, which we can addrss like and array, in 3-d
```
Comment (BrianMichell) [inline examples/dataset_example/src/dataset_example.cc:134]: /// The read returns a SharedArray, which we can address like and array, in 3-d

File: examples/dataset_example/src/dataset_example.cc [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+
+  std::string dataset_path = temp_dir / "zarrs/seismic_3d/";
+
+  auto json_spec = GetToyExample();
+
+  /// A dataset is used to represent a collection of data "on disk".
+  /// The MDIO json schema for this can be found here:
+  ///    mdio/dataset_schema.h
+  auto dataset_future = mdio::Dataset::from_json(json_spec, dataset_path,
+                                                 mdio::constants::kCreateClean);
```
Comment (BrianMichell) [inline examples/dataset_example/src/dataset_example.cc:52]: We should probably leave a comment here about the overwrite potential of using the `kCreateClean` option.

Comment (blasscoc) [reply examples/dataset_example/src/dataset_example.cc:65]: Hi Brian, i've updated with changes.

**Review by BrianMichell — APPROVED** [review]

**Pull request merged**